### ISSUE TITLE
test_tcp_internals.PlatformAssumptionsTestCase fails when we leak fds

### ISSUE THREAD

**Issue by twisted-trac** [twisted/twisted#5301]
|[<img alt="itamarst's avatar" src="https://avatars.githubusercontent.com/u/3266662?s=50" width="50" height="50">](https://github.com/itamarst)| @itamarst reported|
|-|-|
|Trac ID|trac#5301|
|Type|defect|
|Created|2011-10-09 20:29:22Z|
|Branch|https://github.com/twisted/twisted/tree/fd-limit-5301|

If we leak fds, the default chosen socket limit may end up being *below* the actual number of fds, and so the `test_acceptOutOfFiles` will fail before it even can get to any assertions.

Obviously, leaking fds is bad, but sometimes (e.g. gtk) it's not our fault, and we should be resilient in the face of such problems. In particular, the tests in [#7153](https://github.com/twisted/twisted/issues/7153) make the gtk leaking worse on some platforms.

<details><summary>Searchable metadata</summary>

```
trac-id__5301 5301
type__defect defect
reporter__itamar itamar
priority__normal normal
milestone__ 
branch__branches_fd_limit_5301 branches/fd-limit-5301
branch_author__itamarst itamarst
status__closed closed
resolution__fixed fixed
component__core core
keywords__ 
time__1318192162000000 1318192162000000
changetime__1318421377000000 1318421377000000
version__None None
owner__itamar itamar

```
</details>

**Comment by twisted-trac** [comment]
|[<img alt="itamarst's avatar" src="https://avatars.githubusercontent.com/u/3266662?s=50" width="50" height="50">](https://github.com/itamarst)<a name="note_1"></a>|@itamarst commented|
|-|-|

(In [32803]) Branching to 'fd-limit-5301'

**Comment by twisted-trac** [comment]
|[<img alt="itamarst's avatar" src="https://avatars.githubusercontent.com/u/3266662?s=50" width="50" height="50">](https://github.com/itamarst)<a name="note_2"></a>|@itamarst set owner to @exarkun|
|-|-|

exarkun reviewed current version of the fix, and said "The twisted/test/test_tcp_internals.py will fail to set the new rlimit if the number of open files is close to (within 10) the open file limit."

I don't really understand what that means, so assigning to exarkun for clarification.

**Comment by twisted-trac** [comment]
|[<img alt="itamarst's avatar" src="https://avatars.githubusercontent.com/u/3266662?s=50" width="50" height="50">](https://github.com/itamarst)<a name="note_3"></a>|@itamarst commented|
|-|-|

Make the need for input more explicit.

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_4"></a>|@exarkun set owner to @itamarst|
|-|-|

The case I was trying to point out was this one.  

   * The pre-test hard file limit is 256
   * The pre-test open file count is 250

The computed new limit will be 260 and the setrlimit call will fail because this exceeds the hard limit and the tests will likely be running without permissions to exceed the hard limit.

It should be possible for the test to operate with two available file descriptors.  One for the listening port, one for the client connection.  So I think replacing `+ 10` with `+ 2` (and perhaps adding a comment) will make the test as good as we can make it.

**Comment by twisted-trac** [comment]
|[<img alt="itamarst's avatar" src="https://avatars.githubusercontent.com/u/3266662?s=50" width="50" height="50">](https://github.com/itamarst)<a name="note_5"></a>|@itamarst removed owner|
|-|-|

I decided it was simpler (and less fragile in the fact of the known limitations of _listOpenFDs) to just set the soft limit to its highest possible value, i.e. the hard limit.

Buildbot run:
http://buildbot.twistedmatrix.com/boxes-supported?branch=/branches/fd-limit-5301

**Comment by twisted-trac** [comment]
|[<img alt="itamarst's avatar" src="https://avatars.githubusercontent.com/u/3266662?s=50" width="50" height="50">](https://github.com/itamarst)<a name="note_6"></a>|@itamarst set owner to @itamarst|
|-|-|

Which apparently is no good? What?

**Comment by twisted-trac** [comment]
|[<img alt="itamarst's avatar" src="https://avatars.githubusercontent.com/u/3266662?s=50" width="50" height="50">](https://github.com/itamarst)<a name="note_7"></a>|@itamarst removed owner|
|-|-|

Plan (A) (_listOpenFDS() + 2) seems to have worked.

http://buildbot.twistedmatrix.com/boxes-supported?branch=/branches/fd-limit-5301

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_8"></a>|@exarkun set owner to @itamarst|
|-|-|

Thanks.  The new comment in `setUp` says _we might be past the soft limit_, but that's not really true.  Apart from that, looks good.  Merge when that's fixed.

**Comment by twisted-trac** [comment]
|[<img alt="itamarst's avatar" src="https://avatars.githubusercontent.com/u/3266662?s=50" width="50" height="50">](https://github.com/itamarst)<a name="note_9"></a>|@itamarst set status to `closed`|
|-|-|

(In [32823]) Merge fd-limit-5301.

Author: itamar
Review: exarkun
Fixes: [#5301](https://github.com/twisted/twisted/issues/5301)

Make twisted.test.test_tcp_internals more robust in the face of FD leaks (*cough*pygtk*cough*).
